Update i18n recipe (Translate Routes, Add how to hide the default language in the URL) (#4161)
withastro/docs: src/content/docs/en/recipes/i18n.mdx

This example serves each language at its own subpath, e.g. `example.com/en/blog` for English and `example.com/fr/blog` for French.

> If you don't want the default language to show up in the URL like the other languages, you can use the [instructions](/en/recipes/i18n/#hide-default-language-in-the-url) below.
> 

Reviewer: ```suggestion
If you prefer the default language to not be visible in the URL unlike other languages, there are [instructions to hide the default language](/en/recipes/i18n/#hide-default-language-in-the-url) below.

```

    </html>
    ```

### Hide default language in the URL

> 1. Create a directory only for each language that is not the default language. For example, store your default language pages direclty in `pages/` and `fr/` if you also support French:

Reviewer: ```suggestion
1. Create a directory for each language except the default language. For example, store your default language pages directly in `pages/`, and your translated pages in `fr/`:
```

    <FileTree>
    - src/
      - pages/
        - about.astro
        - index.astro
        - **fr/**
          - about.astro
          - index.astro
        - index.astro
    </FileTree>

> 2. Add one line to the `src/i18n/ui.ts` file to possibly toggle the feature:

Reviewer: ```suggestion
2. Add another line to the `src/i18n/ui.ts` file to toggle the feature:
```

    ```ts
    // src/i18n/ui.ts
    export const showDefaultLang = false;
    ```

> 3. Add helper function to the `src/i18n/utils.ts` file to translate paths based on the current language:

Reviewer: ```suggestion
3. Add a helper function to `src/i18n/utils.ts`, to translate paths based on the current language:
```

   ```js
    // src/i18n/utils.ts
    import { ui, defaultLang, showDefaultLang } from './ui';

    export function useTranslatedPath(lang: keyof typeof ui) {
      return function translatePath(path: string, l: string = lang) {
        return !showDefaultLang && l === defaultLang ? path : `/${l}${path}`
      }
    }
    ```

> 4. Import the helper where needed. For example, a nav component might look like:

Reviewer: ```suggestion
4. Import the helper where needed. For example, a `nav` component might look like:
```

    ```astro 
    ---
    // src/components/Nav.astro
    import { getLangFromUrl, useTranslations } from '../i18n/utils';
    
    const lang = getLangFromUrl(Astro.url);
    const t = useTranslations(lang);
    const translatePath = useTranslatedPath(lang);
    ---
    <ul>
        <li>
            <a href={translatePath('/home/')}>
              {t('nav.home')}
            </a>
        </li>
        <li>
            <a href={translatePath('/about/')}>
              {t('nav.about')}
            </a>
        </li>
        <li>
            <a href="https://twitter.com/astrodotbuild">
              {t('nav.twitter')}
            </a>
        </li>
    </ul>
    ```

> 5. Helper could also be used to translate paths for specific language. For example, when users switch between languages:

Reviewer: ```suggestion
5. The helper function can also be used to translate paths for a specific language. For example, when users switch between languages:
```

    ```astro
    ---
    // src/components/LanguagePicker.astro
    import { languages } from '../i18n/ui';
    ---

    <ul>
      {Object.entries(languages).map(([lang, label]) => (
        <li>
          <a href={translatePath('/', lang)}>{label}</a>
        </li>
      ))}
    </ul>
    ```

### Translate Routes

> Translate the routes of your pages for each language.
> 
> 
> 1. Add route mappings to the `src/i18n/ui.ts` file:

Reviewer: ```suggestion
Translate the routes of your pages for each language.

1. Add route mappings to `src/i18n/ui.ts`:
```

 ```ts
 // src/i18n/ui.ts
 export const routes = {
   de: {
    'services': 'leistungen',
   },
   fr: {
    'services': 'prestations-de-service',
   },
  }
    ```

> 2. Update the `useTranslatedPath` helper function in `src/i18n/utils.ts` to add some more route translation logic.

Reviewer: ```suggestion
2. Update the `useTranslatedPath` helper function in `src/i18n/utils.ts` to add router translation logic.
```

  ```js
  // src/i18n/utils.ts
  import { ui, defaultLang, showDefaultLang, routes } from './ui';

  export function useTranslatedPath(lang: keyof typeof ui) {
    return function translatePath(path: string, l: string = lang) {
      const pathName = path.replaceAll('/', '')
      const hasTranslation = defaultLang !== l && routes[l] !== undefined && routes[l][pathName] !== undefined
      const translatedPath = hasTranslation ? '/' + routes[l][pathName] : path
  
      return !showDefaultLang && l === defaultLang ? translatedPath : `/${l}${translatedPath}`
    }
  }
  ```

> 3. Create a helper function to get the route if it exists based on the current URL in `src/i18n/utils.ts`:

Reviewer: ```suggestion
3. Create a helper function to get the route, if it exists based on the current URL, in `src/i18n/utils.ts`:
```

  ```js
  // src/i18n/utils.ts
  import { ui, defaultLang, showDefaultLang, routes } from './ui';
  
  export function getRouteFromUrl(url: URL): string | undefined {
    const pathname = new URL(url).pathname
    const parts = pathname?.split('/')
    const path = parts.pop() || parts.pop()
  
    if (path === undefined) {
      return undefined
    }
    
    const currentLang = getLangFromUrl(url);
  
    if (defaultLang === currentLang) {
      const route = Object.values(routes)[0];
      return route[path] !== undefined ? route[path] : undefined
    }
    
    const getKeyByValue = (obj: Record<string, string>, value: string): string | undefined  => {
        return Object.keys(obj).find((key) => obj[key] === value)
    }
  
    const reversedKey = getKeyByValue(routes[currentLang], path)
  
    if (reversedKey !== undefined) {
      return reversedKey
    }
  
    return undefined
  }
  ```

> 4. Helper could be used to get a translated route. For example, when users switch languages, they don't have to come back from the home page. However, if no route is defined, you will be taken to the home page

Reviewer: ```suggestion
4. The helper function can be used to get a translated route. For example, when no translated route is defined, the user will be redirected to the home page:
```
